src/Platform/Microsoft.Testing.Extensions.TrxReport/Hashing/XxHash128.cs

```diff
@@ -0,0 +1,423 @@
+        {
+            Hash128 hash = HashToHash128(source, seed);
+            WriteBigEndian128(hash, destination);
+            bytesWritten = HashLengthInBytes;
+            return true;
+        }
+
+        bytesWritten = 0;
+        return false;
+    }
+
+    private static Hash128 HashToHash128(byte[] source, long seed = 0)
+    {
+        uint length = (uint)source.Length;
+        fixed (byte* sourcePtr = source)
+        {
+            return HashToHash128(sourcePtr, length, seed);
+        }
+    }
+
+    private static Hash128 HashToHash128(byte* sourcePtr, uint length, long seed = 0)
+    {
+        if (length <= 16)
+        {
+            return HashLength0To16(sourcePtr, length, (ulong)seed);
+        }
+
+        if (length <= 128)
+        {
+            return HashLength17To128(sourcePtr, length, (ulong)seed);
+        }
+
+#pragma warning disable IDE0046 // Convert to conditional expression
+        if (length <= MidSizeMaxBytes)
+        {
+            return HashLength129To240(sourcePtr, length, (ulong)seed);
+        }
+#pragma warning restore IDE0046 // Convert to conditional expression
+
+        return HashLengthOver240(sourcePtr, length, (ulong)seed);
+    }
+
+    private Hash128 GetCurrentHashAsHash128()
+    {
+        Hash128 current;
+
+        if (_state.TotalLength > MidSizeMaxBytes)
+        {
+            // Digest on a local copy to ensure the accumulators remain unaltered.
+            ulong* accumulators = stackalloc ulong[AccumulatorCount];
+            CopyAccumulators(ref _state, accumulators);
+
+            fixed (byte* secret = _state.Secret)
+            {
+                DigestLong(ref _state, accumulators, secret);
+                current = new Hash128(
+                    low64: MergeAccumulators(accumulators, secret + SecretMergeAccsStartBytes, _state.TotalLength * Prime64_1),
+                    high64: MergeAccumulators(accumulators, secret + SecretLengthBytes - (AccumulatorCount * sizeof(ulong)) - SecretMergeAccsStartBytes, ~(_state.TotalLength * Prime64_2)));
+            }
+        }
+        else
+        {
+            fixed (byte* buffer = _state.Buffer)
+            {
+                current = HashToHash128(buffer, (uint)(int)_state.TotalLength, (long)_state.Seed);
+            }
+        }
+
+        return current;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static void WriteBigEndian128(in Hash128 hash, byte[] destination)
+    {
+        ulong low = hash.Low64;
+        ulong high = hash.High64;
+        if (BitConverter.IsLittleEndian)
+        {
+            low = BinaryPrimitives.ReverseEndianness(low);
+            high = BinaryPrimitives.ReverseEndianness(high);
+        }
+
+#if NET
+        ref byte dest0 = ref MemoryMarshal.GetReference(destination);
+        Unsafe.WriteUnaligned(ref dest0, high);
+        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref dest0, new IntPtr(sizeof(ulong))), low);
```

Comment:
This is the original logic from dotnet/runtime.

On TFMs where we don't have these APIs, I'm doing it as:

```csharp
        fixed (byte* dest0 = destination)
        {
            *(ulong*)(void*)dest0 = high;
            *((ulong*)(void*)dest0 + 1) = low;
        }
```

@jkotas Could you please help us verify if the semantics are still always correct here?

Reply:
We want to avoid forcing dependencies on MTP users in most cases.